### ISSUE TITLE
C interface for the rust core

### ISSUE THREAD

**Issue by leccyworld** [opendp/opendp#235]
I am exploring feasibility to use your library on an embedded arm platform (aarch64). 
-Cross compile the rust core with the correct toolchain
-Use the C interface to invoke the core APIs

Could you please let me know if this is possible at all?

**Comment by Shoeboxam** [comment]
I don't have a lot of knowledge about embedded platforms, but aarch64 is listed as a [rustc supported platform](https://doc.rust-lang.org/nightly/rustc/platform-support.html).
You can use [this page](https://rust-lang.github.io/rustup/overrides.html#default-toolchain) for info on how to set the toolchain.
We do provide a C interface via the opendp-ffi crate. I'm curious how much space you have on the embedded device. It might make sense to add some feature flags to the ffi crate to cut down the binary size.

**Comment by leccyworld** [comment]
Hello, Thanks for the info. Is there any example on how to use the opendp-ffi?

**Comment by leccyworld** [comment]
> I don't have a lot of knowledge about embedded platforms, but aarch64 is listed as a [rustc supported platform](https://doc.rust-lang.org/nightly/rustc/platform-support.html).
> You can use [this page](https://rust-lang.github.io/rustup/overrides.html#default-toolchain) for info on how to set the toolchain.

Cross compilation is failing with below error:
error: failed to run custom build command for `gmp-mpfr-sys v1.2.4`

Caused by:
  process didn't exit successfully: `/home/sgw/opendp/opendp/rust/target/release/build/gmp-mpfr-sys-08e966bab8b9263e/build-script-build` (exit status: 101)
  --- stdout
  cargo:warning=Cross compilation not supported! Use the `force-cross` feature to cross compile anyway. (This warning will become an error in version 1.3.)
  cargo:rerun-if-env-changed=GMP_MPFR_SYS_CACHE

**Comment by Shoeboxam** [comment]
Try adding the `--no-default-features` flag to help you get things bootstrapped. This drops the gmp/mpfr dependency, and along with it the continuous distribution mechanisms. The geometric mechanism should remain accessible. 

What kinds of mechanisms are you wanting to use? For example, I have a randomized response algorithm in a PR that you might be useful for embedded systems.

There are a couple different routes you could take with ffi:

The more general approach is the FFI crate, which does dispatch and monomorphization over a collection of relevant types for each generic argument, as well as type erasure on many of the inputs and outputs. This is really useful for general purpose bindings like in python, but may (or may not) be too heavy for your embedded is case. The python bindings are relevant as they use the FFI crate. I could talk more about the core api.

If you only have a few specific data types and computation chains you need to work with, then you could write a thin ffi crate that exposes just the functionality you need. I think we should have an example crate that demonstrates how this would be done anyways. I'll open a PR with this tomorrow.

**Comment by leccyworld** [comment]
I had to take out openssl as well as that was causing issues with finding "limits.h". 
Now I have libopendp_ffi.so loading on my target chip (ARM64 based). I tried that by linking the library to a simple C program but I am not sure if any of the functionalities are available through C like APIs. 
I could run python on  my platform as well. Is there any example I can try with python and libopendp_ffi?

**Comment by Shoeboxam** [comment]
[Here is an example of the python bindings (which uses the FFI crate).](https://github.com/opendp/opendp/blob/main/python/example/demo.ipynb)

When it comes to using the FFI crate directly from C, I'll point you to a few links to demonstrate usage of the Laplace mechanism. 
Any function that starts with `extern "C"` should be callable on libopendp_ffi.so. [This is the function for creating a laplace measurement.](https://github.com/opendp/opendp/blob/main/rust/opendp-ffi/src/meas/laplace.rs#L16-L31). 

You can call it from C by passing a void pointer to a 32-bit or 64-bit scale float, and a c_char pointer like "AllDomain<f64>" (for a 64-bit laplace mechanism on scalars) or "VectorDomain<AllDomain<f32>>" (for a 32-bit laplace mechanism on vectors). 

You'll notice that constructor returns an FfiResult. We've configured FfiResult to have a C representation. [You can see the layout of this C data structure based on how we interpret it in python here.](https://github.com/opendp/opendp/blob/main/python/src/opendp/_lib.py#L82-L101). Once you unwrap this in C you've got a pointer to a measurement.

At this point there are quite a few useful extern functions in `/rust/opendp-ffi/src/core/mod.rs`.
[Free errors](https://github.com/opendp/opendp/blob/main/rust/opendp-ffi/src/core/mod.rs#L183-L187)
[Evaluate the privacy relation](https://github.com/opendp/opendp/blob/main/rust/opendp-ffi/src/core/mod.rs#L202-L213)
[Invoke measurement](https://github.com/opendp/opendp/blob/main/rust/opendp-ffi/src/core/mod.rs#L215-L220)
[Free measurement](https://github.com/opendp/opendp/blob/main/rust/opendp-ffi/src/core/mod.rs#L222-L225)
Etc.

You'll notice that you need to pass an AnyObject when invoking the measurement, and you get an AnyObject back. 
[You can load data up into an AnyObject with this.](https://github.com/opendp/opendp/blob/main/rust/opendp-ffi/src/data/mod.rs#L81-L82)
[You can unload data from an AnyObject with this.](https://github.com/opendp/opendp/blob/main/rust/opendp-ffi/src/data/mod.rs#L187-L188)
Notice that these translate between an [FfiSlice](https://github.com/opendp/opendp/blob/main/rust/opendp-ffi/src/core/mod.rs#L13-L17) and and AnyObject. 

In summary, the steps to invoke the function of a laplace measurement over vectors: 
1. Create an FfiSlice, where you pack your array behind the pointer and include the length of the array.
2. Call out to `opendp_data_slice_as_object` to convert the FfiSlice into an AnyObject.
3. Invoke the measurement by passing the measurement pointer and AnyObject pointer into `opendp_core__measurement_invoke`. 
4. Unpack the AnyObject into an FfiSlice via `opendp_data__object_as_slice`.

**Comment by Shoeboxam** [comment]
About approach 2, here's an example crate for wiring your own ffi up directly, on your own:
https://github.com/opendp/opendp/commit/8561d7e57e960eb72fffa9f24e2dbe54bb6084bb

**Comment by leccyworld** [comment]
Hi , Thanks for the suggestions. I was able to load libopendp.so and run the inbuilt rust tests on my platform. So I think first approach should be OK for the evaluation on my arm platform in terms of memory and performance. 
I am trying to run libopendp_ffi.so now. Can you put together a simple C example on how to use the APIs? A real world use case would be good. Something like  The farmer/rabbit example from google DP library? Not sure if that's possible with opendp core

**Comment by pdurbin** [comment]
I wasn't familiar with the farmer example from Google's DP library but here it is (and it's pretty fun!): https://github.com/google/differential-privacy/blob/main/examples/cc/README.md

> There are around 200 animals at Farmer Fred's zoo. Every day, Farmer Fred feeds
the animals as many carrots as they desire. The animals record how many carrots
they have eaten per day. For this particular day, the number of carrots eaten
can be seen in [animals_and_carrots.csv](https://github.com/google/differential-privacy/blob/main/examples/cc/animals_and_carrots.csv).
> 
> At the end of each day, Farmer Fred often asks aggregate question about how many
carrots everyone ate. For example, he wants to know how many carrots are eaten
each day, so he knows how many to order the next day. The animals are fearful
that Fred will use the data against their best interest. For example, Fred could
get rid of the animals who eat the most carrots!
> 
> To protect themselves, the animals decide to use the C++ Differential Privacy
library to aggregate their data before reporting it to Fred. This way, the
animals can control the risk that Fred will identify individuals' data while
maintaining an adequate level of accuracy so that Fred can continue to run the
zoo effectively.

**Comment by leccyworld** [comment]
yes thats the one :-)

**Comment by raprasad** [comment]
2/29/24 - keep open to revisit in the future

**Comment by raprasad** [comment]
Note: this functionality exists but not documented; e.g. extending Rust into C or vice versa
